Import victor's layout system 🍷

nox opened the pull request:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/servo/servo/24165)
<!-- Reviewable:end -->

nox committed:
Kill fragments

nox committed:
Remove some more code

nox committed:
Rename RecalcStyleAndConstructFlows to RecalcStyle

It doesn't construct flows anymore.

nox committed:
Import victor's layout system 🍷

highfive commented:
Heads up! This PR modifies the following files:
 * @emilio: components/style/values/generics/length.rs, components/style/style_adjuster.rs, components/layout/model.rs, components/style/values/computed/length.rs, components/style/values/specified/box.rs

highfive commented:
<img src="http://www.joshmatthews.net/warning.svg" alt="warning" height=20> **Warning** <img src="http://www.joshmatthews.net/warning.svg" alt="warning" height=20>

* These commits modify style and layout code, but no tests are modified. Please consider adding a test!

SimonSapin committed:
Import files from Victor

https://github.com/SimonSapin/victor/tree/fdb11f3e87f6d2d59170d10169fa6deb94e53b94

emilio commented on components/style/values/specified/box.rs:
```diff
@@ -235,6 +273,7 @@ impl Display {
     #[inline]
     fn from3(outside: DisplayOutside, inside: DisplayInside, list_item: bool) -> Self {
         let inside = match inside {
+            #[cfg(any(feature = "servo-layout-2013", feature = "gecko"))]
```
Why not leaving it computing `flow` to `inline` / `block` as everywhere else?

nox replied:
This is compiled out because I removed a lot of variants from `DisplayInside`, most notably `Block` and `Inline`.

emilio commented on components/style/values/specified/box.rs:
```diff
@@ -352,12 +393,14 @@ impl Display {
         match self.outside() {
             DisplayOutside::Inline => {
                 let inside = match self.inside() {
+                    #[cfg(any(feature = "servo-layout-2013", feature = "gecko"))]
```
`#[cfg]`ing this out is wrong, why?

nox replied:
I can't keep it, I removed `DisplayInside::Block` and `DisplayInside::Inline`. Those are hacks to make layout supposedly easier, but it's just hindering layout 2020.

nox replied:
I added back a blockification which makes it store `DisplayInside::FlowRoot` with `servo-layout-2020`.

nox replied:
Err, I meant I made it store `DisplayInline::Flow`, sorry.

SimonSapin commented on components/layout_2020/primitives.rs:
```diff
@@ -1,28 +1,29 @@
-#[derive(Copy, Clone, PartialEq)]
-pub struct RGBA(pub f32, pub f32, pub f32, pub f32);
+// #[derive(Copy, Clone, PartialEq)]
+// pub struct RGBA(pub f32, pub f32, pub f32, pub f32);
```
Let’s remove the parts of this file that are commented out (we’re unlikely to use them as-is) and move the rest to `geom.rs`.

`primitives.rs` was first created for Victor’s PDF backend.

SimonSapin commented on components/layout_2020/style_ext.rs:
```diff
@@ -0,0 +1,127 @@
+        if packed_display == PackedDisplay::None {
+            return Self::None;
+        }
+        // FIXME(nox): Support display: contents.
```
Isn’t `contents` handled here the same as `none`?

SimonSapin commented on components/layout_2020/flow/construct.rs:
```diff
@@ -134,7 +152,7 @@ impl BlockContainer {
                 return (container, builder.contains_floats);
             }
-            builder.end_ongoing_inline_formatting_context();
+            // builder.end_ongoing_inline_formatting_context();
```
Maybe add FIXME comments near those pieces of commented-out code, so we can find them later and distinguish them more easily from "normal" text comments?

SimonSapin commented:
r+ with nits

Thanks @nox!

nox commented:
@bors-servo r=SimonSapin

bors-servo commented:
:pushpin: Commit 4444c5a has been approved by `SimonSapin`

<!-- @bors-servo r=SimonSapin 4444c5a2adfec909e165497ca331c09e02ee54cc -->

bors-servo commented:
:sunny: Test successful - [linux-rel-css](http://build.servo.org/builders/linux-rel-css/builds/13366), [linux-rel-wpt](http://build.servo.org/builders/linux-rel-wpt/builds/13428), [status-taskcluster](https://tools.taskcluster.net/task-group-inspector/#/RjIBPe_cT7-ZD2YHFURFgQ)
Approved by: SimonSapin
Pushing 86ad84198ef0f53b12e0eecfacd9edbfa0d27a1a to master...
